src/cargo/core/compiler/custom_build.rs

```diff
@@ -359,7 +365,19 @@ fn build_work<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>) -> CargoRes
                 },
                 true,
             )
-            .chain_err(|| format!("failed to run custom build command for `{}`", pkg_name))?;
+            .chain_err(|| format!("failed to run custom build command for `{}`", pkg_name));
+
+        if output.is_err() {
+            insert_warnings_in_build_output(
+                build_script_outputs,
+                id,
+                metadata_hash,
+                warnings_in_case_of_panic,
+            );
+            return output.map(|_| ());
```

Comment:
Instead of using a closure here, perhaps do something like:
```rust
if let Err(e) = output {
    …
    return Err(e);
}
```

Just seems a slightly clearer to me.

Reply:
Ah, this was actually the first version I wrote, but changed it to this one to avoid needless wrapping/unwrapping. Will change back